### PR TITLE
Allow nullable types in collections

### LINKED ISSUE #163: Support nullable types in collections
RealmResulsts should support null values

depends on https://github.com/realm/realm-dart/issues/178

### PR DESCRIPTION
**Pull request by nielsenko**
Ultimately this is about supporting nullable **primitive** elements on realm lists, ie.
```dart
@RealmModel()
class _Player {
  @PrimaryKey()
  late String name;
  _Game? game;
  final scoresByRound = <int?>[]; // null means player didn't finish
}

@RealmModel()
class _Game {
  final winnerByRound = <_Player>[]; // I intended to support <_Player?> here as well, but alas core doesn't support it
  int get rounds => winnerByRound.length;
}
```
In the process I refactored how the generator lays out the schema, so it is now all done compile time with const constructable objects, so the above will generate `Player`:
```dart
class Player extends _Player with RealmEntityMixin, RealmObjectMixin {
  Player(
    String name, {
    Game? game,
    Iterable<int?> scoresByRound = const [],
  }) {
    _nameProperty.setValue(this, name);
    _gameProperty.setValue(this, game);
    _scoresByRoundProperty.setValue(this, RealmList<int?>(scoresByRound));
  }

  Player._();

  static const _nameProperty = ValueProperty<String>(
    'name',
    RealmPropertyType.string,
    primaryKey: true,
  );
  @override
  String get name => _nameProperty.getValue(this);
  @override
  set name(String value) => throw RealmUnsupportedSetError();

  static const _gameProperty = ObjectProperty<Game>('game', 'Game');
  @override
  Game? get game => _gameProperty.getValue(this);
  @override
  set game(covariant Game? value) => _gameProperty.setValue(this, value);

  static const _scoresByRoundProperty =
      ListProperty<int?>('scoresByRound', RealmPropertyType.int);
  @override
  RealmList<int?> get scoresByRound => _scoresByRoundProperty.getValue(this);
  @override
  set scoresByRound(covariant RealmList<int?> value) =>
      throw RealmUnsupportedSetError();

  @override
  Stream<RealmObjectChanges<Player>> get changes =>
      RealmObjectMixin.getChanges(this);

  static const schema = SchemaObject<Player>(
    Player._,
    'Player',
    {
      'name': _nameProperty,
      'game': _gameProperty,
      'scoresByRound': _scoresByRoundProperty,
    },
    _nameProperty,
  );
  @override
  SchemaObject get instanceSchema => schema;
}
```
Notice how `schema` is const constructed, and the strongly typed property types.

Dynamic realms, use `_DynamicSchemaProperty` and benefits by allowing me to drop `getList<T>` for just `get<List<T>>`, ie. not special property access needed for lists.

Resolves #163

### CODE REVIEW

**Review by desistefanova — COMMENTED** [review]
Maybe it is by a reason, but it looks more like refactoring than only supporting nullable in collections with primitives. It would be nice if the hierarchy of classes and mixins was simpler. Maybe after the discussion it will become clearer to me.

File: test/list_test.dart [hunk 1]
```diff
@@ -699,4 +699,71 @@ Future<void> main([List<String>? args]) async {
...
+    checkResult([
+      bob,
+      alice,
+      bob
+    ], {
+      alice: [1, 3, 2],
+      bob: [2, null, 3],
+      carol: [null, 1, 1]
+    });
+  });
```
[the remark this answers is not in the document]
Comment (nielsenko) [reply test/list_test.dart:793]: This does not test all the cases tested by the original test case?

**Review by blagoev — CHANGES_REQUESTED** [review]
I think we are doing too much in this PR. We definitely need to have a discussion about the changes around the generator. I think we are putting too much responsibility into the generator which was designed to be lightweight and something we plan to replace in the future.

**Comment by nielsenko** [comment]
> I think we are doing too much in this PR. We definitely need to have a discussion about the changes around the generator. I think we are putting too much responsibility into the generator which was designed to be lightweight and something we plan to replace in the future.

There is no more logic in the generator than before, it is just as fast as it has always been. It just lays out the generated code in a smarter way. It allows the runtime to skip work, because more is known compile time. I don't see how this in anyway will make it more difficult to transition to static meta-programming once it lands.

**Comment by nielsenko** [comment]
This is superseded by #902 and #903 which splits the feature implementation from the refactoring

**Comment by Shreedhar73** [comment]
@nielsenko Hello.
Are Nullable types in Collections are only supported for local database ? 

When i try to have a field of Type List<String?> it is throwing sync error.
Invalid schema change (UPLOAD): sync does not support collections of nullable primitives unless using the mixed type  . 

Is it the expected behavour or I am missing something here?

**Comment by nirinchev** [comment]
That is correct. The server team does plan to eventually add support but it hasn't been a highly requested feature and thus we don't have any timeframe for when we'll get to it. If it's important for your app, definitely reach out to your AE to clarify your requirements.

**Comment by Shreedhar73** [comment]
@nirinchev Thanks for Clarification.

**Pull request closed without merging**